[JSC] Use simde for find16,32,64 functions

Constellation opened the pull request:
#### 91262be0cb68c36ed872eeedeed6db12b821e211
<pre>
[JSC] Use simde for find16,32,64 functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=273849">https://bugs.webkit.org/show_bug.cgi?id=273849</a>
<a href="https://rdar.apple.com/127701822">rdar://127701822</a>

Reviewed by Mark Lam.

This patch starts using simde for find16 / find32 / find64 implementations.
And we also use a slightly different approach so that we no longer need to access to out-of-ASAN region (but still within page boundary).
So we can remove SUPPRESS_ASAN. And we also use the same code for x64 too.

* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find16AlignedImpl): Deleted.
(WTF::find32AlignedImpl): Deleted.
(WTF::find64AlignedImpl): Deleted.
* Source/WTF/wtf/text/StringCommon.h:

Canonical link: <a href="https://commits.webkit.org/278641@main">https://commits.webkit.org/278641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d320df7dbd179bb957b2e1fc8b90bcdcd1ac313

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3535 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/54457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/1890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1564 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/54457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53299 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/44133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/54457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/1389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/44539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/1462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56053 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50702 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/26310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/27558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/44200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/28439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/62936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7443 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/62936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WTF/wtf/SIMDHelpers.h:
```diff
@@ -116,13 +166,61 @@ ALWAYS_INLINE std::optional<uint8_t> findFirstNonZeroIndex(simde_uint16x8_t valu
 #else
+    //  value                     |c|c|c|C|c|C|c|c| (c is character, C is matching character)
+    //  eq with charactersVector  |0|0|0|X|0|X|0|0| (X is all-1)
+    //  Reduce to uint8x8_t       |0|0|0|X|0|X|0|0| => reinterpret it as uint64_t. If it is non-zero, matching character exists.
```
This is misleading.  Let's just say this instead:
// Incoming value is a comparison result, where each vector element is either all 1s or 0s.

Constellation replied:
Changed.

ddegazio commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -609,101 +607,50 @@ ALWAYS_INLINE const uint16_t* find16(const uint16_t* pointer, uint16_t character
     // If the pointer is unaligned to 16bit access, then SIMD implementation does not work. But ARM64 allows unaligned access.
     // Fallback to a simple implementation. We also use it for smaller memory where length is less than 16.
     constexpr size_t thresholdLength = 32;
```
Maybe we should decrease the threshold? I suppose it's not part of your changes but I'd be surprised if 16-31 byte text isn't profitable to vectorize, since our alternative is a byte-by-byte loop.

Constellation replied:
Yeah, for now, this patch is keeping this threshold. But we could make this smaller after this.

ddegazio reviewed (approved):
LGTM, with one small question/suggestion.

MenloDorian commented on Source/WTF/wtf/SIMDHelpers.h:
```diff
@@ -116,13 +166,61 @@ ALWAYS_INLINE std::optional<uint8_t> findFirstNonZeroIndex(simde_uint16x8_t valu
+    //
+    // Example
+    //     mask        |0|0|0|X|0|X|0|0| (X is all-one)
+    //     not-mask    |X|X|X|0|X|0|X|X|
```
Let's use the term "value" instead of "mask" in the comment lines so that it matches the incoming argument name.  Otherwise, it can be confusing.

Constellation replied:
Changed.

Achierius commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -609,101 +607,50 @@ ALWAYS_INLINE const uint16_t* find16(const uint16_t* pointer, uint16_t character
+        constexpr size_t stride = 16 / sizeof(UnsignedType);
+        static_assert(stride <= thresholdLength);
+        auto charactersVector = SIMD::splat(character);
+        for (; cursor + (stride - 1) < end; cursor += stride) {
+            auto value = SIMD::load(cursor);
+            auto mask = SIMD::equal(value, charactersVector);
+            if (auto index = SIMD::findFirstNonZeroIndex(mask))
+                return cursor + index.value();
+        }
```
Previously this code first walked up to an alignment boundary -- is that hidden inside of load/equal/findFirstNonZeroIndex now?
If so then the comment on 607/608 seems out of date

Constellation replied:
Previously, we are using alignment boundary, and scanning the memory regardless of the last remaining out-of-range memory (since it fits in page boundary). But this new approach is just scanning with stride and the last one is scanned with the adjusted starting point.

MenloDorian commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -597,10 +597,8 @@ ALWAYS_INLINE const uint8_t* find8(const uint8_t* pointer, uint8_t character, si
+ALWAYS_INLINE const UnsignedType* findImpl(const UnsignedType* pointer, UnsignedType character, size_t length)
 {
     // We take `size_t` length instead of `unsigned` because,
     // 1. It is aligned to memchr.
```
The `// 2. It allows us to use find16 for 4GB~ vectors, which can be used in JSC ArrayBuffer (4GB wasm memory).` comment below this is stale, and should be updated.  "find16" ==> "findImpl".

Constellation replied:
Removed.

MenloDorian commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -609,101 +607,50 @@ ALWAYS_INLINE const uint16_t* find16(const uint16_t* pointer, uint16_t character
     // If the pointer is unaligned to 16bit access, then SIMD implementation does not work. But ARM64 allows unaligned access.
     // Fallback to a simple implementation. We also use it for smaller memory where length is less than 16.
```
This comment also looks stale, especially the "16bit access" part.  Maybe should be "character width"? 

Ditto for the "length is less than 16." part since your `thresholdLength` is 32.  Maybe should just use thresholdLength instead of 16 here?

Constellation replied:
Changed. We no longer need this comment. Just mentioning to threshold.

MenloDorian reviewed (approved):
r=me with comments fixed.

Constellation committed:
[JSC] Use simde for find16,32,64 functions
https://bugs.webkit.org/show_bug.cgi?id=273849
rdar://127701822

Reviewed by Mark Lam.

This patch starts using simde for find16 / find32 / find64 implementations.
And we also use a slightly different approach so that we no longer need to access to out-of-ASAN region (but still within page boundary).
So we can remove SUPPRESS_ASAN. And we also use the same code for x64 too.

* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find16AlignedImpl): Deleted.
(WTF::find32AlignedImpl): Deleted.
(WTF::find64AlignedImpl): Deleted.
* Source/WTF/wtf/text/StringCommon.h:

Canonical link: https://commits.webkit.org/278641@main
